src/lib/reducers/apps.js

```diff
@@ -1,68 +1,73 @@
 import stack from '../stack'
+import { ForbiddenException } from '../exceptions'
 
 // constants
-const SET_APPS = 'SET_APPS'
+const RECEIVE_APP_LIST = 'RECEIVE_APP_LIST'
+const RECEIVE_APP_LIST_FORBIDDEN = 'RECEIVE_APP_LIST_FORBIDDEN'
 const SET_INFOS = 'SET_INFOS'
 const EXCLUDES = ['settings', 'onboarding']
 const CATEGORIES = ['cozy', 'partners', 'ptnb']
 
-// reducers
-export const getApps = state => state.apps
+// selectors
+export const getApps = state => state.apps && state.apps.data
 export const getAppsFiltered = state => {
-  if (!state.apps) return []
-  return state.apps.filter(app =>
+  if (!state.apps || !state.apps.data) return []
+  return state.apps.data.filter(app =>
     app.name !== state.appName || app.editor !== state.editor
   )
 }
+export const isAppListForbidden = state => state.apps ? state.apps.forbidden : false
 
 // actions
-const setApps = apps => ({ type: SET_APPS, apps })
+const receiveAppList = apps => ({ type: RECEIVE_APP_LIST, apps })
+const receiveAppListForbidden = () => ({ type: RECEIVE_APP_LIST_FORBIDDEN })
 export const setInfos = (appName, editor) => ({ type: SET_INFOS, appName, editor })
 
 // actions async
 export const fetchApps = () => async dispatch => {
-  let apps
   try {
-    apps = await Promise.all((await stack.get.apps())
-      .filter(app => !EXCLUDES.includes(app.attributes.slug))
-      .map(async app => {
-        // TODO load only one time icons
-        const srcIcon = await stack.get.icon(app.links.icon)
-        const icon = srcIcon
-          ? {
-            src: srcIcon,
-            cached: true
-          }
-          : undefined
-        return {
-          editor: app.attributes.editor,
-          name: app.attributes.name,
-          slug: app.attributes.slug,
-          href: app.links.related,
-          category: CATEGORIES.includes(app.attributes.category)
-            ? app.attributes.category
-            : 'others',
-          icon
+    const rawAppList = await stack.get.apps()
+    const apps = rawAppList.filter(app => !EXCLUDES.includes(app.attributes.slug))
+    // TODO load only one time icons
+    const icons = await Promise.all(apps.map(app => stack.get.icon(app.links.icon)))
+    const appsWithIcons = apps.map((app, idx) => ({
+      editor: app.attributes.editor,
+      name: app.attributes.name,
+      slug: app.attributes.slug,
+      href: app.links.related,
+      category: CATEGORIES.includes(app.attributes.category)
+        ? app.attributes.category
+        : 'others',
+      icon: icons[idx]
+        ? {
+          src: icons[idx],
+          cached: true
         }
-      })
-    )
-    await dispatch(setApps(apps))
+        : undefined
+    }))
+    dispatch(receiveAppList(appsWithIcons))
   } catch (e) {
-    console.warn(e.message ? e.message : e)
+    if (e.constructor === ForbiddenException) {
+      dispatch(receiveAppListForbidden())
+    } else {
+      console.warn(e.message ? e.message : e)
+    }
```

Comment:
We shall wrap by a try/catch the only instruction that need it. Maybe by extracting this instruction:

Instead of
```js
try {
  cont app = getApp()
  app.doSomething()
} catch (Exception e) {
  dispatch(noApp())
}
```
we can do
```js
const getAppWithExceptionHandling = () => {
  try {
    return getApp()
  } catch (Exception e) {
    return null
  }
}

const app = getAppWithExceptionHandling()
if(app === null) {
  dispatch(noApp())
  return
}
app.doSomething()
```

What do you think?

Reply:
The `stack.get.apps()` call is not the only instruction needing to be try/catched, there are also the `stack.get.icon()` calls, that's why I put everything in the try/catch (and besides that, I need to dispatch a very specific action for a very specific error, so your suggestion wouldn't work). I thought splitting this code even more could make it less readable, and I already made it more clearer than it was imho.